hw1/task1/get_stats.py

```diff
@@ -0,0 +1,140 @@
+    def return_function(x):
+        try:
+            x = type_to_convert(x)
+        except ValueError:
+            return return_if_exception
+        else:
+            return x
+    return return_function
+
+
+def read_value_data(path):
+    data = pd.read_csv(path, error_bad_lines=False)
+    data.columns = np.char.lower(np.array(data.columns, dtype=str))
+
+    data['lpep_dropoff_datetime'] = data['lpep_dropoff_datetime'].apply(convert_type_function(np.datetime64))
+    data['lpep_pickup_datetime'] = data['lpep_pickup_datetime'].apply(convert_type_function(np.datetime64))
+    data['trip_distance'] = data['trip_distance'].apply(convert_type_function(float))
+    data['total_amount'] = data['total_amount'].apply(convert_type_function(float))
+    data['passenger_count'] = data['passenger_count'].apply(convert_type_function(int))
+    return data
+
+
+def general_stats(path=None, data=None):
+    if data is None and path is not None:
+        data = read_value_data(path)
+    elif path is None and data is not None:
+        data = data
+    else:
+        return None
+
+    gen_stat = {}
+    required_columns = ['lpep_pickup_datetime',
+                        'lpep_dropoff_datetime',
+                        'total_amount',
+                        'passenger_count',
+                        'trip_distance']
+
+    invalid_rows = np.any(data[required_columns].isna(), axis=1).sum()
+
+    clean_data = data[required_columns].dropna()
+
+    gen_stat['mean_cost'] = clean_data['total_amount'].mean()
+    clean_data['trip_duration'] = clean_data['lpep_dropoff_datetime'] - clean_data['lpep_pickup_datetime']
+    longest = clean_data['trip_duration'].max()
+    gen_stat['longest_ride'] = longest
+
+    pickup_times = pd.DataFrame(clean_data['lpep_pickup_datetime'])
+    if len(pickup_times) != 0:
+        pickup_times.index = pickup_times['lpep_pickup_datetime']
+        ten_minutes_resample = pickup_times['lpep_pickup_datetime'].resample('10T').count()
+        max_count_start = ten_minutes_resample.idxmax()
+        max_count = ten_minutes_resample.max()
+        max_count_end = max_count_start + np.timedelta64(10, 'm')
+        gen_stat['max_count'] = max_count
+        gen_stat['max_count_start'] = max_count_start
+        gen_stat['max_count_end'] = max_count_end
+
+    gen_stat['invalid_rows'] = invalid_rows
+    gen_stat['count'] = clean_data.shape[0]
+    gen_stat = pd.DataFrame(gen_stat, index=[0])
+    return gen_stat
+
+
+def missing_dates(path=None, data=None):
+    if data is None and path is not None:
+        data = read_value_data(path)
+    elif path is None and data is not None:
+        data = data
+    else:
+        return None
+    required_columns = ['lpep_pickup_datetime',
+                        'lpep_dropoff_datetime',
+                        'passenger_count',
+                        'trip_distance',
+                        'total_amount']
+
+    clean_data = data[required_columns].dropna()
```

Comment:
This code without any changes is repeated throughout the program 4 times.
This should signal you, that it could be re-written as separate function.
Moreover, due to this the code reads data 4 times, which is extra IO cost. It's better to read the file before calculating any statistic and then perform calculations.
This will make your code clean and decoupled. I.e. functions, that calculate statistic will be separated from functions, that read the file, writing output file should also go separately